### PR TITLE
feat(locators): Add support for regex in cssContainingText

### PR DESCRIPTION
**Pull request by joeheyming**
In order to get this working, I had to serialize the regex before
we send it over the wire to the browser.

Since there is no standard way to do this, I took guidance from
a stackoverflow answer, where they call toString on the regex.
Then, on the browser, you use a regex to extract out the text in between
/someregex/

The hard part is to also extract out the modifiers, like i for ignore case.

Anyhow, this approach seems to work, and I have tests to prove it.

This is related to #1360

### CODE REVIEW

**Comment by joeheyming** [comment]
I looked at the continuous test failure, and the failures are unrelated to my changes.
I touched cssContainingText.
The test that failed, protractor/spec/ts/basic/element_spec.ts:177:3)
has nothing to do with cssContaining text.  It is trying to fill out a form

**Comment by joeheyming** [comment]
Not sure what to do with this:

running: node scripts/interactive_tests/with_base_url.js
.
/home/ubuntu/protractor/node_modules/q/q.js:155
                throw e;
                ^
Did not start interactive server in 10s. Server startup output: [19:50:49] I/local - Starting selenium standalone server...
[19:50:51] I/local - Selenium standalone server started at http://172.17.197.126:44518/wd/hub


>>> fail: Error: expecting exit code: 0, actual: 1
running: node node_modules/jasmine/bin/jasmine.js JASMINE_CONFIG_PATH=scripts/unit_test.json

**Comment by joeheyming** [comment]
Hey so... Ping.... echo... echo... echo... echoo........

**Comment by wswebcreation** [comment]
Hi @joeheyming 

Sorry for not reacting. I'll add someone from the core team to review this PR.

File: lib/clientsidescripts.js [hunk 1]
```diff
@@ -676,20 +676,26 @@ functions.findByPartialButtonText = function(searchText, using) {
...
+ * @param {string} searchText The exact text to match.  Can be a serialized regex.
  * @param {Element} using The scope of the search.
  *
  * @return {Array.<Element>} An array of matching elements.
  */
 functions.findByCssContainingText = function(cssSelector, searchText, using) {
   using = using || document;
 
+  if (searchText.indexOf('__REGEXP__') === 0) {
+    var match = searchText.split('__REGEXP__')[1].match(/\/(.*)\/(.*)?/);
```
Comment (qiyigg) [inline lib/clientsidescripts.js]: I am not sure enough whether this match method works well here

Would you mind adding another test that using a regex with no regex flags, we can be more convinced of this change

File: lib/clientsidescripts.js [hunk 2]
```diff
@@ -676,20 +676,26 @@ functions.findByPartialButtonText = function(searchText, using) {
  * Find elements by css selector and textual content.
  *
  * @param {string} cssSelector The css selector to match.
- * @param {string} searchText The exact text to match.
+ * @param {string} searchText The exact text to match.  Can be a serialized regex.
```
Comment (qiyigg) [inline lib/clientsidescripts.js]: Would it be better if we say: searchText The exact text to match or a serialized regex.

File: lib/clientsidescripts.js [hunk 3]
```diff
@@ -676,20 +676,26 @@ functions.findByPartialButtonText = function(searchText, using) {
...
+  }
   var elements = using.querySelectorAll(cssSelector);
   var matches = [];
   for (var i = 0; i < elements.length; ++i) {
     var element = elements[i];
     var elementText = element.textContent || element.innerText || '';
-    if (elementText.indexOf(searchText) > -1) {
+    
+    if ((searchText instanceof RegExp && searchText.test(elementText)) ||
+        elementText.indexOf(searchText) > -1) {
```
Comment (qiyigg) [inline lib/clientsidescripts.js]: My understanding is "elementText.indexOf(searchText) > -1" should only be applied to "searchText is not a RegExp"
For now, if elementText doesn't satisfy searchText regex, "elementText.indexOf(searchText) > -1" will still be called

File: lib/clientsidescripts.js (hunk 2 above)
Comment (joeheyming) [reply lib/clientsidescripts.js]: sure

File: lib/clientsidescripts.js (hunk 3 above)
Comment (joeheyming) [reply lib/clientsidescripts.js]: There are no side-effects to calling indexOf.  The way || works is that javascript will take the first true value.
So I don't think indexOf will be called.

Here is an example I tried out:
x = () => { console.log('here'); return true; }
y = () { console.log('there'); return false; }

z = x() || y()
here
true

File: lib/clientsidescripts.js (hunk 1 above)
Comment (joeheyming) [reply lib/clientsidescripts.js]: ok, will do

**Comment by joeheyming** [comment]
I did a force push to my branch on my fork.  Does github do well with that?

File: lib/clientsidescripts.js (hunk 3 above)
Comment (joeheyming) [reply lib/clientsidescripts.js]: Ok, I tried to address this issue.  Let me know if you like the updated code.

**Comment by joeheyming** [comment]
The last CircleCi failure was a pip install failure.  Nothing to see here

**Pull request merged**